src/LogicItLab/Salesforce/MapperBundle/WsdlValidator.php

```diff
@@ -0,0 +1,192 @@
+
+    /** @var string */
+    private $relativePath;
+
+    /** @var string */
+    private $wsdlPath;
+
+    /**
+     * @param AnnotationReader $annotationReader
+     * @param string $srcPath
+     * @param string $relativePath
+     * @param string $wsdlPath
+     * @codeCoverageIgnore
+     */
+    public function __construct(AnnotationReader $annotationReader, string $srcPath, string $relativePath, string $wsdlPath)
+    {
+        $this->annotationReader = $annotationReader;
+        $this->srcPath = $srcPath;
+        $this->relativePath = $relativePath;
+        $this->wsdlPath = $wsdlPath;
+    }
+
+    /**
+     * @return array
+     * @throws FileHandlerNotInitializedException
+     * @throws \ReflectionException
+     */
+    public function retrieveMissingFields(): array
+    {
+        $missingFields = [];
+
+        $fileHandler = fopen($this->wsdlPath, "r");
+        $this->initializeFileHandler($fileHandler);
+
+        foreach ($this->getAllClassAnnotations() as $annotation) {
+            $objectName = $annotation['object']->name;
+            $fieldNames = $this->getFieldNames($annotation['fields']);
+
+            if (!$objectName || !$fieldNames) {
+                continue;
+            }
+
+            if ($this->hasObject($fileHandler, $objectName) === false) {
+                $missingFields[$objectName] [] = "entire object";
+                continue;
+            }
+
+            foreach ($fieldNames as $fieldName) {
+                if ($this->hasField($fileHandler, $fieldName) === false) {
+                    $missingFields[$objectName] [] = $fieldName;
+                }
+            }
+        }
+
+        return $missingFields;
+    }
+
+    private function getAllClassAnnotations(): array
+    {
+        $annotations = array_map(function ($className) {
+            return $this->annotationReader->getSalesforceProperties($className);
+        }, $this->getAllClassNames());
+
+        usort($annotations, function ($a, $b) {
+            return strcmp($a['object']->name, $b['object']->name);
+        });
+
+        return $annotations;
+    }
+
+    private function getAllClassNames(): array
+    {
+        $classNames = [];
+        $AllFiles = Finder::create()->files()->in($this->srcPath.$this->relativePath)->name('*.php');
+        foreach ($AllFiles as $file) {
+            $realPath = $file->getRealpath();
+            $fileName = str_replace($this->srcPath, '', $realPath);
+            $className = str_replace('.php', '', $fileName);
+            if(strpos($this->srcPath, 'test') !== false) {
+                $className = "Tests/$className";
+            }
+            $classNames[] = str_replace('/', '\\', $className);
+        }
+
+        $classNames = array_filter($classNames, function ($element) {
+            return class_exists($element);
+        });
+
+        sort($classNames);
+        return $classNames;
+    }
+
+    private function hasObject($fileHandler, $objectName): bool
```

Comment:
If I understood correctly you look though a file every time you call this function. Does the file name change or only the object name?

Reply:
I'd suggest reading the file once, put it in a variable and use regular expressions instead. The wsdls are ~3mb, there is no reason to read it line by line. And as a bonus we'll lose this over-complicaded pieces of code.

Reply:
file does not change and it is read just once, check is done while file is read. I don't know how to use regular expression for this. Is not just about look if a string is present in the file but I need to check that is contained in the correct object. Don't know if I explained myself. I tried using simple xml before this, but at my eyes looked easier this approach, the structure of the result xml was complicate.

Reply:
Here's something I've tried to explain. Works fine. Adjust it as you wish 👍 

```php
<?php

class SomeClass
{
    /** @var string */
    private $wsdlContents;

    private function getWsdlContents(): string
    {
        if(!$this->wsdlContents) {
            $this->wsdlContents = file_put_contents('main_wsdl_file.wsdl');
        }

        return $this->wsdlContents;
    }

    private function getObject(string $objectName): ?string
    {
        preg_match("/<complexType name=\"$objectName\">([\w\W]*)<\/complexType>/", $this->getWsdlContents(), $matches);

        return $matches[0] ?? null;
    }

    private function hasField(string $objectName, string $fieldName): bool
    {
        return strpos($this->getObject($objectName), sprintf('<element name="%s"', $fieldName));
    }
}
```